mempool/src/pool/store.rs

```diff
@@ -0,0 +1,397 @@
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::collections::HashMap;
+
+use common::chain::tokens::OutputValue;
+use common::chain::transaction::Transaction;
+use common::chain::OutPoint;
+use common::primitives::amount::Amount;
+use common::primitives::id::WithId;
+use common::primitives::Id;
+use common::primitives::Idable;
+use common::primitives::H256;
+
+use logging::log;
+
+use utils::newtype;
+
+use crate::error::Error;
+use crate::error::TxValidationError;
+
+use super::Ancestors;
+use super::Conflicts;
+use super::Descendants;
+use super::Time;
+
+newtype! {
+    #[derive(Debug, PartialEq, Eq, Ord, PartialOrd)]
+    pub(super) struct DescendantScore(Amount);
+}
+
+#[derive(Debug)]
+pub(super) struct MempoolStore {
+    // This is the "main" data structure storing Mempool entries. All other structures in the
+    // MempoolStore contain ids (hashes) of entries, sorted according to some order of interest.
+    pub(super) txs_by_id: HashMap<H256, TxMempoolEntry>,
```

Comment:
Somewhat unrelated to the present PR, but could the key here also be `Id<Transaction>` rather than `H256`?

Reply:
Hmm... it could. In the past I ran into issues trying to use `Id<Transaction>` as the key to a hash map. But seems to be possible now. Will open another PR for this